### DISCUSSION TITLE
broker.shared_subscription_strategy hash的疑问

### DISCUSSION THREAD

**Discussion by yuezhongxin** [emqx/emqx#4554] [Q&A] [answered]
配置设置了：
```
broker.shared_subscription_strategy=hash
```

然后有两个订阅者（$queue//sys/#），然后我发了消息，始终只有一个接受到了，按照hash的解释，不应该是谁发送的谁接受到吗？

**Comment by tigercl** [comment]
@yuezhongxin It just means that messages from the same client will be forwarded to a fixed subscriber under the condition that the number of subscriptions remains the same.

**Comment by yuezhongxin** [comment]
哦哦，那和sticky是不是差不多？只是hash在订阅者变化后，再计算clientID选取是吧？后面订阅者不变，就一直是这个接受到。

那如何实现谁（ClientID）发送的谁接受到的呢？

**Comment by yuezhongxin** [comment]
我们的场景是，客户端发送消息，然后同步等待终端回复的消息，现在是广播的形式，比如客户端有很多个，也会接受到，就很不合理。

想做到某个客户端发送，然后只有这个客户端接受到回复的消息，其他客户端接受不到。

**Comment by tigercl** [comment]
@yuezhongxin You can set a certain level of the topic as your client identifier, and then use this topic as the topic of the reply message.

**Comment by yuezhongxin** [comment]
有没有相关文档？不是特别懂，是topic里面加标识吗？

**Comment by zmstone** [comment, marked as answer]
4.3里增加了源topic 的hash:
https://docs.emqx.io/en/broker/v4.3/configuration/configuration.html#broker-shared-subscription-strategy
所以如果订阅者都 subscribe 到 `prefix/#`
然后发布者发布到 `prefix/clientID`
就可以实现类似按照发布者clientID的方式来选择订阅者了

另外，如果订阅者数量发生变化，映射关系就会改变，这个是没法避免的。
